lib/plugins/aws/package/compile/events/stream/index.js

```diff
@@ -247,81 +217,40 @@ class AwsCompileStreamEvents {
-                    }
-                  }
-                  if (
-                    typeof event.stream.destinations.onFailure.arn === 'string' &&
-                    !event.stream.destinations.onFailure.arn.startsWith('arn:')
-                  ) {
-                    throw new this.serverless.classes.Error(
-                      this.resolveInvalidDestinationPropertyErrorMessage(functionName, 'arn')
-                    );
-                  }
-                  OnFailureDestinationArn = event.stream.destinations.onFailure.arn;
-                } else if (typeof event.stream.destinations.onFailure === 'string') {
-                  if (!event.stream.destinations.onFailure.startsWith('arn:')) {
-                    throw new this.serverless.classes.Error(
-                      this.resolveInvalidDestinationPropertyErrorMessage(functionName, 'arn')
-                    );
-                  }
-                  OnFailureDestinationArn = event.stream.destinations.onFailure;
-                } else {
-                  throw new this.serverless.classes.Error(
-                    this.resolveInvalidDestinationPropertyErrorMessage(functionName, 'arn')
-                  );
-                }
+              let OnFailureDestinationArn;
 
-                const destinationType =
-                  event.stream.destinations.onFailure.type || OnFailureDestinationArn.split(':')[2];
-                // add on failure destination ARNs to PolicyDocument statements
-                if (destinationType === 'sns') {
-                  onFailureSnsStatement.Resource.push(OnFailureDestinationArn);
-                } else if (destinationType === 'sqs') {
-                  onFailureSqsStatement.Resource.push(OnFailureDestinationArn);
-                } else {
+              if (typeof event.stream.destinations.onFailure === 'object') {
+                if (
+                  typeof event.stream.destinations.onFailure.arn !== 'string' &&
+                  !event.stream.destinations.onFailure.type
+                ) {
                   const errorMessage = [
-                    `Stream event of function '${functionName}' had unsupported destination type of`,
-                    ` '${streamType}'. Valid stream event source types include 'sns' and`,
-                    " 'sqs'. Please check the docs for more info.",
+                    `Missing "type" property for on failure destination in function "${functionName}"`,
+                    ' If the "arn" property on a destination is a complex type (such as Fn::GetAtt)',
+                    ' then a "type" must be provided for the destination, either "sns" or,',
+                    ' "sqs". Please check the docs for more info.',
                   ].join('');
-                  throw new this.serverless.classes.Properties.Policies[0].PolicyDocument.Error(
-                    errorMessage
-                  );
+                  throw new this.serverless.classes.Error(errorMessage);
```

Comment:
I think we can remove that validation error as well (?)

Reply:
This error message is related to my comment above : since type is only required when arn is not a string, and such validation cannot be written in JSON schema to my knowledge, keeping the error is the only way I found around that.

We could make type required in all case, but this would mean :
- existing serverless configuration relying on arn as a string will have to add this property
- we should remove computing of type from manipulation the arn string, to avoid inconsistency between the type computed from arn string and the type added by the user in the configuration

WDYT ?